### ISSUE TITLE
Error on conditional initialization of Final class variables

### ISSUE THREAD

**Issue by nelsyeung** [python/mypy#10736]
**Bug Report**

Conditional initialization of a Final class variable is reported as a reassignment error on elif/else branches.

**To Reproduce**

`tmp.py`:

```python
from typing import Final, Optional


class Foo:
    a: Final[Optional[int]]

    def __init__(self, v: int) -> None:
        if v == 0:
            self.a = None
        else:
            self.a = v  # error: Cannot assign to final attribute "a"  [misc]
```

**Expected Behavior**

No errors on the first correct assignment of a final variable regardless of branching.

**Actual Behavior**

```sh
$ mypy tmp.py
tmp.py:11: error: Cannot assign to final attribute "a"  [misc]
Found 1 error in 1 file (checked 1 source file)
```
Reported `error: Cannot assign to final attribute "a"  [misc]` during the else clause, even though it's only the first assignment of the variable.

**Your Environment**

- Mypy version used: 0.902
- Mypy command-line flags: `mypy tmp.py`
- Mypy configuration options from `pyproject.toml` (and other config files):
 
    ```toml
    [tool.mypy]
    show_error_codes = true
    ```
- Python version used: 3.9.5
- Operating system and version: macOS 11.4

**Related**
* Pyright has already fixed this issue: https://github.com/microsoft/pyright/issues/1917

**Comment by randolf-scholz** [comment]
Match-case statements are also affected: https://mypy-play.net/?mypy=master&python=3.11&gist=f0b81218a489585562acd1d8468c2f26

```python
from typing import Final

class Foo:
    bar: Final[tuple[int, ...]]
    
    def __init__(self, bar: None | int | tuple[int, ...]) -> None:
        match bar:
            case None:
                self.bar = (-1,)
            case int():
                self.bar = (bar,)  # ✘ Cannot assign to final attribute "bar"  [misc]
            case tuple():
                self.bar = bar     # ✘ Cannot assign to final attribute "bar"  [misc]
```

**Comment by merc1031** [comment]
I have a similar issue / conundrum.
Is there anyway to (forward) declare the type of a variable that will be used in and exhaustive match?

https://mypy-play.net/?mypy=master&python=3.11&gist=10f92cdd6e414525c63d25ae251f9afa
```
def __init__(self, foo: None | int | tuple[int, ...]) -> tuple[int, ...]:
    bar: Final[tuple[int, ...]]  # ✘ Final name must be initialized with a value  [misc]
    match foo:
        case None:
            bar = (-1,)  # ✘ Cannot assign to final attribute "bar"  [misc]
        case int():
            bar = (foo,)  # ✘ Cannot assign to final attribute "bar"  [misc]
        case tuple():
            bar = foo  # ✘ Cannot assign to final attribute "bar"  [misc]
        case tuple():
            bar = foo  # ✘ Cannot assign to final attribute "bar"  [misc]

    return bar
```

**Comment by randolf-scholz** [comment]
@merc1031 Try

```python
from typing import Final

class Foo:
    bar: Final[tuple[int, ...]]

    def __init__(self, foo: None | int | tuple[int, ...]) -> None:
        bar: tuple[int, ...]
        match foo:
            case None:
                bar = (-1,)
            case int():
                bar = (foo,)
            case tuple():
                bar = foo
            case tuple():
                bar = foo
    
        self.bar = bar
```

**Comment by merc1031** [comment]
@randolf-scholz 
If  you notice my use case is without the class / instance variable. Only a function.

In a larger context it would be to protect from someone modifying bar before the end of the function

https://mypy-play.net/?mypy=master&python=3.11&gist=372aa35d3e4f02d35aabe38a3b709473

```
from typing import Final


def __init__(self, foo: None | int | tuple[int, ...]) -> tuple[tuple[int, ...], bool]:
    bar: Final[tuple[int, ...]]
    match foo:
        case None:
            bar = (-1,)
        case int():
            bar = (foo,)
        case tuple():
            bar = foo
        case tuple():
            bar = foo
    # more code happens here
    # protect from bar being reassigned
    return bar, True
 ```

**Comment by Avasam** [comment]
I feel like there's 2 issues here.
1. is the assignement to a Final attribute in a class' `__init__`. Duplicate of https://github.com/python/mypy/issues/8982 
2. is the conditional assignement to a Final variable when all code paths include said assignement. And a ternary is not possible. For example:
```py
if pytorch_is_available:
    import torch

    pytorch_gpu_is_available: Final = torch.cuda.is_available()
else:
    # We only assign once
    pytorch_gpu_is_available: Final = False  # type: ignore[misc]
```
